src/librustdoc/doctest.rs

```diff
@@ -372,30 +371,47 @@ fn wrapped_rustc_command(rustc_wrappers: &[PathBuf], rustc_binary: &Path) -> Com
     command
 }
 
-fn run_test(
-    test: &str,
-    crate_name: &str,
+pub(crate) struct DocTestInfo {
+    line_offset: usize,
     line: usize,
-    rustdoc_options: IndividualTestOptions,
+    path: PathBuf,
+}
+
+fn build_test_dir(outdir: &Arc<DirState>, is_multiple_tests: bool) -> PathBuf {
+    // Make sure we emit well-formed executable names for our target.
+    let is_perm_dir = matches!(**outdir, DirState::Perm(..));
+    let out_dir = outdir.path();
+    let out_dir = if is_multiple_tests && is_perm_dir {
+        // If this a "multiple tests" case and we generate it into a non temporary directory, we
+        // want to put it into the parent instead.
+        out_dir.parent().unwrap_or(out_dir)
```

Comment:
It can be improved by only pushing the test id if it's not a multiple test indeed. It cannot be done in `make_test` because we don't know yet if this is a multiple test.